rotkehlchen/chain/base/modules/echo/decoder.py

```diff
@@ -0,0 +1,199 @@
+import logging
+from collections.abc import Callable
+from typing import TYPE_CHECKING, Any
+
+from rotkehlchen.assets.asset import EvmToken
+from rotkehlchen.chain.base.modules.echo.constants import (
+    CPT_ECHO,
+    DEAL_ABI,
+    DEAL_FUNDED,
+    ECHO_CPT_DETAILS,
+    FEE_PAID,
+    FUNDER_DEREGISTERED,
+    FUNDER_REGISTRY,
+    FUNDING_CONDUIT,
+    POOL_REFUNDED,
+)
+from rotkehlchen.chain.ethereum.utils import asset_normalized_value
+from rotkehlchen.chain.evm.decoding.interfaces import DecoderInterface
+from rotkehlchen.chain.evm.decoding.structures import (
+    DEFAULT_DECODING_OUTPUT,
+    ActionItem,
+    DecoderContext,
+    DecodingOutput,
+)
+from rotkehlchen.chain.evm.decoding.types import CounterpartyDetails
+from rotkehlchen.constants.resolver import evm_address_to_identifier
+from rotkehlchen.history.events.structures.types import HistoryEventSubType, HistoryEventType
+from rotkehlchen.logging import RotkehlchenLogsAdapter
+from rotkehlchen.types import ChecksumEvmAddress, EvmTokenKind, EvmTransaction
+from rotkehlchen.utils.misc import bytes_to_address
+
+if TYPE_CHECKING:
+    from rotkehlchen.chain.evm.structures import EvmTxReceiptLog
+    from rotkehlchen.history.events.structures.evm_event import EvmEvent
+
+logger = logging.getLogger(__name__)
+log = RotkehlchenLogsAdapter(logger)
+
+
+class EchoDecoder(DecoderInterface):
+
+    def _decode_fee_paid(self, context: DecoderContext) -> DecodingOutput:
+        """Transform transfer event to Echo fee event if Echo charged on-chain fee"""
+        if (
+            context.tx_log.topics[0] != FEE_PAID or
+            not self.base.is_tracked(user_address := bytes_to_address(context.tx_log.topics[1]))
+        ):
+            return DEFAULT_DECODING_OUTPUT
+
+        if fee_amount_from_input := int.from_bytes(context.transaction.input_data[36:68]):
+            return DEFAULT_DECODING_OUTPUT
+
+        deal_address = bytes_to_address(context.transaction.input_data[68:100])
+        raw_token_address = self.evm_inquirer.call_contract(
+                contract_address=deal_address,
+                abi=DEAL_ABI,
+                method_name='token',
+        )
+        token = EvmToken(evm_address_to_identifier(
```

Comment:
this should also be a get_or_create_evm_token as Yabir mentioned in the other function below